pkg/virt-controller/services/template.go

```diff
@@ -561,6 +567,12 @@ func (t *templateService) RenderLaunchManifest(vmi *v1.VirtualMachineInstance) (
 		nodeSelector[k] = v
 
 	}
+	if featuregates.CPUNodeDiscoveryEnabled() {
+		if vmi.Spec.Domain.CPU != nil && vmi.Spec.Domain.CPU.Model != "" {
+			nodeSelector[NFD_CPU_FAMILY_PREFIX+vmi.Spec.Domain.CPU.Model] = "true"
```

Comment:
do we want (or need) to sanitize the CPU.Model to make sure the label is well formed?
Example (not sure it applies): is the label name case sensitive? The CPU.Model? What if they are and the name doesn't match?

Furthermore: do we have a common database of cpu models?
I mean: is the cpu label essentially a glorified free string or do we want to validate the labels?

Reply:
> do we want (or need) to sanitize the CPU.Model to make sure the label is well formed?
Example (not sure it applies): is the label name case sensitive? The CPU.Model? What if they are and the name doesn't match?

The cpu model is provided from VM/VMI spec (CPU.Model) and will be used as a label in the node selector we add to virt-launcher:
1. If the model name doesn't comply with the [labels and selectors syntax](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) it will fail during the creation of virt-launcher with a proper error message. This syntax seems to be compatible with [libvirt cpu models](https://github.com/libvirt/libvirt/tree/master/src/cpu_map) (correct me if im wrong)
2. If the model name comply with labels syntax but somehow was written incorrectly (doesn't match  existing cpu families) it wont be scheduled : 
`Message:               0/1 nodes are available: 1 node(s) didn't match node selector.`
labels are case sensitive so if there is a mismatch -  no scheduling happens. 
btw, i see in that case an added value to the existing behaviour we had so far. non existing model names wont be mapped to libvirt's domain spec (converter stage after virt-launcher is up)

All of the above is under the assumption that the node-feature-discovery plugin will set the correct model names on the nodes.

> Furthermore: do we have a common database of cpu models?
I mean: is the cpu label essentially a glorified free string or do we want to validate the labels?

No, we do not have a common database of cpu models nor we check cpu models against a list (in webhook validations for example). 
Conceptually we can maybe gather a list of all existing cpu models from all nodes using the node-feature-discovery plugin and then use it for verification, but  it seems robust.

Reply:
I think we should check (on VM level) for syntax compatibility and we might want to limit the values to whatever libvirt supports in the future (the value is also passed to libvirt xml). But, we do not have access to libvirt at the VM level and so there is no easy way to do this atm.

Reply:
> Conceptually we can maybe gather a list of all existing cpu models from all nodes using the node-feature-discovery plugin and then use it for verification,

I wouldn't do this.


But what we need is a coordination between the CPU models we use in VMI, which are libvirt cpu models. And the models exposed by nfd.

I can imagine  that the names differ between what libvirt calls cpu families, and what nfd calls cpu families.

How do we solve this issue?

Reply:
Fabian, nfd does not expose any families atm. We will be writing the flags + cpuid -> families logic.

Also, for scheduling reasons, the node has to expose all supported families labels (the cpu one + all older that are compatible), because the POD only has one and should run on newer machines as well.

Reply:
+1

Reply:
But if you were talking about validation only, then explicit list of supported families is of course better.

Reply:
> We will be writing the flags + cpuid -> families logic.

Hopefully we can leverage libvirt here?



> Also, for scheduling reasons, the node has to expose all supported families labels (the cpu one + all older that are compatible)

Absolutely



> But if you were talking about validation only, then explicit list of supported families is of course better.

Well, not _only_, but mainly yes.